packages/ember-glimmer/lib/utils/references.js

```diff
@@ -52,6 +52,27 @@ export class UpdatableReference extends RootReference {
   }
 }
 
+export class GetHelperReference {
+  constructor(args) {
+    this.positionalArgs = args.positional;
+  }
+
+  isDirty() { return true; }
+
+  value() {
+    let objRef = this.positionalArgs.at(0).value(),
+      pathRef = this.positionalArgs.at(1).value();
+
+    return objRef ? get(objRef, pathRef) : objRef;
+  }
+
+  get(propertyKey) {
+    throw new Error("get() is not implemented");
```

Comment:
In general, a `Reference` is an abstract data type with the following interface:

```
interface Reference<T> {
  value(): T;
}
```

A reference represents a "computation" (similar to a "stream"), and calling `.value()` will return the current value of that computation. For example, a reference representing `{{foo}}` will return the current value of the `foo` property in the template's context. This gives you a stable value/wrapper (hence the name "reference") to hold on to that you can query in the future to re-render the template with the latest value as needed.

A `PathReference` is an extension to the system:

```
interface PathReference<T> extends Reference<T> {
  get(key: string): PathReference<any>;
}
```

Implementing the `PathReference` interface means that the `value()`s of the computation supports a conceptual `get` operation. For the purpose of understanding this, you can apply the `Ember.get` semantics here – i.e. the `value()`s returned by this computation makes sense as input to the first argument of `Ember.get`.

Alternatively, you can also think about the `get` semantics in terms of the templates. Anything `foo` that can be used in a `{{foo.bar}}` position must implement `PathReference`. In this case, `{{foo.bar}}` will take the `{{foo}}` reference as input and call `.get("bar")` on it to produce another stable reference (another `PathReference, actually) representing `{{foo.bar}}`.

In practice, almost anything that can be used in a template needs to implement `PathReference`, because in handlebars almost anything can be used in that position. In addition to @chadhietala's example, there are a few other ways that a `get` helper can get into that position: `{{(get foo bar).baz}}`, `{{#with (get foo bar) as |gotBar|}} {{gotBar.baz}} {{/with}}`, etc.

In this case, the reference you are implementing is representing the result of the `(get foo bar)` computation over time, where foo and bar themselves are also `PathReferences`. The `get(path)` method (confusingly named in this context) should produce a reference representing the `(get foo bar).path` computation.

Putting everything together, I think something like this would work:

```
...
toReference(args) {
  let sourceReference = args.positional.at(0); // foo in (get foo bar)
  let propertyPathReference = args.positional.at(1); // bar in (get foo bar)

  if (isConst(propertyPathReference)) {
    // (get foo "bar"), so `propertyPathReference` is a reference whose value cannot change ever
    // this is pretty pointless, you could have just typed `{{foo.bar}}`
    return sourceReference.get(propertyPathReference.value());
  } else {
    return GetHelperReference(sourceReference, propertyPathReference);
  }
}
...

class GetHelperReference {
  // ...mostly following your implementation

  get(path) {
    new PropertyReference(this, path);
  }
}
```